### ISSUE TITLE
Support Extension nodes

### ISSUE THREAD

**Issue by binarycow** [zspitz/ExpressionTreeToString#34]
Requesting support for extension nodes, i.e., expressions with a `NodeType` of `ExpressionType.Extension`

**Comment by zspitz** [comment]
How would the third-party parsing of a given extension expression be used by the library?

I suppose ideally there would be some kind of plugin mechanism (#12). Then a registered formatter class (inheriting from [`WriterBase.cs`](https://github.com/zspitz/ExpressionTreeToString/blob/master/Library/WriterBase.cs)) could parse the extension node, overriding a (to-be-written) `WriteExtensionNode` method.

I would be interested in hearing your thoughts on this, and if you could elaborate on your specific use case.

**Comment by binarycow** [comment]
First, my recommendation...

I currently call `expression.ToString("C#");`.  I would like to call `expression.ToString("C#", new CustomFormatter());`.

`CustomFormatter` would implement the following:

```
interface ICustomExpressionTreeFormatter
{
     public bool TryFormatExtension(Expression expression /* plus whatever other parameters I would need */);
     public bool TryFormatBinary(BinaryExpression expression /* plus whatever other parameters I would need */);
     public bool TryFormatBlock(BlockExpression expression /* plus whatever other parameters I would need */);
     public bool TryFormatLambda(LambdaExpression expression /* plus whatever other parameters I would need */);
     /* Other methods, as needed */
}
```

Your code, would do something like this:

```
private void WriteBinaryExpression(BinaryExpression expression)
{
    if(this.customFormatter != null)
    {
        if(this.customFormatter.TryFormatBinary(expression /* ...... */))
            return;
    }
    /* ....... */
}
```
----

My use case:

----

I am writing expression trees for a dynamically typed language.  The expression tree will be compiled, for performance reasons.  Since types are not known at compile time, and C# is a statically typed language, then all of my expression trees would be stuck using Reflection, or dynamic, and removing the benefits of compiling the expression tree.

My first iteration at solving this problem was to generate an expression tree with everything being `typeof(object)`.  When presented with a set of input parameters, I can then use an `ExpressionVisitor` to walk the tree, and replacing the "object" expressions with strongly typed ones.

Take, for example, this expression:

```
myVariable.SomeProperty[alpha] + somethingElse;
```

Without any types, we cannot make an expression tree.  So - I created a class...

```
// Implementation isn't necessary, methods are never actually called.
class static ReflectionHelper {
   public static object PropertyOrField(object target, string propertyOrFieldName) => throw new Exception();
   public static object ArrayIndex(object target, object index) => throw new Exception();
   public static object Add(object left, object right) => throw new Exception();
}
```

Now, my expression tree becomes this:

```
var myVariable = Expression.Parameter(typeof(object), "myVariable");
var somethingElse = Expression.Parameter(typeof(object), "somethingElse");
var alpha = Expression.Parameter(typeof(object), "alpha");

var body = Expression.Call(typeof(ReflectionHelper), "Add", null,
   Expression.Call(typeof(ReflectionHelper), "ArrayIndex", null,
      Expression.Call(typeof(ReflectionHelper), "PropertyOrField", null,
         myVariable,
         Expression.Constant("SomeProperty")
      ),
      alpha     
   ),
   somethingElse
);

var lambda = Expression.Lambda<Func<object, object, object, object>>(body, new ParameterExpression[] { myVariable, somethingElse, alpha });
```

Store that lambda into a wrapper class.  When I call the `Invoke` method of that wrapper class, I provide argument values and names.  The `Invoke` method re-orders the argument values based on parameter names.  Then, it takes the types of the arguments, and determines if it has a compiled delegate cached for those types.  If so, it executes it.  If not....

An `ExpressionVisitor` will run over the tree.  Based on my new-found knowledge of the types of those parameters, it will replace the `typeof(object)` `ParameterExpression` with new ones, that are strongly typed, and detect the `MethodCallExpression`s that invoke static methods on `ReflectionHelper`, and replace them with new strongly typed expressions.

----

The problem here, is that my expression trees semantically represent method calls, instead of array index, etc.  Additionally, it means I have to process every single `MethodCallExpression`, determine if its a call to `ReflectionHelper`, etc.

I thought that a simpler method may be to create what I call a `DeferredExpression`.  These expressions all have a return type of `DeferredResult`.

So, `DeferredBinaryExpression` has a `Left`, `Right`, and `NodeType` property, just like the regular `BinaryExpression` property.  It is semantically, exactly the same.  The `ExpressionVisitor` will now have an easier time working with it.  `DeferredBinaryExpression` is semantically the exact same thing as a `BinaryExpression`, and is **directly** convertible to `BinaryExpression` - as long as all `DeferredExpressions` have been "resolved".  So, as long as I do a depth-first traversal, and replace any `ParameterExpression` with their new types, I should be able to convert the entire tree to standard C# expressions.

The problem is, that this may end up being a multi-step process, especially when you consider that lambdas can be passed as arguments to other lambdas, etc.  So, step 1 may be to replace all of the `ParameterExpression`, then step 2 may be to replace the expressions within any "leaf" blocks (i.e., blocks that don't use have any lambda parameters)., converting "Deferred" lambdas into concrete type lambdas, etc.  Rinse and repeat until the entire tree is statically typed.  

Because this is a multi-step process, during debugging, I'd like to do a `.ToString("C#")` and see what the intermediate results are, after each step.

Once all steps are complete, and I have removed all "Deferred" nodes, I can compile lambda, and use it any time the user presents that same set of variable types.

**Comment by zspitz** [comment]
(I'm sorry it's taken me some time to respond; I've been carefully considering how to do this.)

Passing `"C#"` to the `ToString` extension method is simply a way of indicating which formatter to use. I think it appropriate to add an additional overload which would accept a custom formatter instead of the formatter name as a string.

`WriterBase.cs` could serve as the base class for any formatter. It has the following low-level methods:

- `WriteNode`
- `Write`
- `Indent`
- `Dedent`
- `WriteEOL`

Currently, `WriterBase` also has `abstract` methods for parsing and rendering specific expression tree types - `WriteBinary` for `BinaryExpression`, `WriteLambda` for `LambdaExpression`. These abstract methods are not always needed (see [TextualTreeFormatter](https://github.com/zspitz/ExpressionTreeToString/blob/master/Library/TextualTreeFormatter.cs#L87), where the implementation of all these methods simply forwards the expression to a central `WriteTextualNode` method), so I would separate these into a different base class in the hierarchy, like so:

```text
WriterBase                                         low-level writing methods, and WriteNode
    TextualTreeFormatter *
    ObjectNotationFormatter *
    BuiltinsFormatter                             forces handling of built-in expression types
        FactoryMethodsFormatter *
        CodeWriter
            CSharpCodeWriter *
            VisualBasicCodeWriter *
```

With this hierarchy, which level to inherit from depends on your needs:

- If you don't care about the built-in expression types at all, inherit from `WriterBase`. When a given node is passed to `WriteNode`, the implementation would forward the child expressions of the node to recursively call `WriteNode`.
- If you want to handle the built-in expression types in a new way unrelated to existing formatters -- e.g. XML or JSON data extracted from the expression tree -- inherit from `BuiltinsFormatter`, which forces an implemetation to parse each type in `System.Linq.Expressions`. The point is to avoid what appears to have happened in the built-in `.ToString` and `DebugView` -- some nodes have C#-like representations, but other nodes have a generic representation, e.g. `Convert(...)` for conversion nodes.
- If you want to extend the C# formatter, you can do so, optionally overriding the base methods as needed; for example, in order to handle various extension expression types.

In short, I plan on doing the following:

- [ ] Split the current `WriterBase` into `WriterBase` and `BuiltinsFormatter`
    - [ ] Simplify implementations of `TextualTreeFormatter` and `ObjectNotationFormatter`
- [ ] Add `ParseExtension` to `BuiltinsFormatter`
- [ ] Separate `Parse...` methods from `Write....` methods
- [ ] Enable the `ToString` extension method to take a `WriterBase`-derived instance

Once this is complete, you'll be able to write a class that inherits from `WriterBase` and pass an instance of that class to `.ToString`.

Will this work for you?

**Comment by zspitz** [comment]
(NB. I'm not sure if this helps, but the Dynamic Language Runtime (DLR) has a mechanism for recompiling the delegate at a given call site. AFAICT, it works something like this:

Let's say you have the following source code expression:
```
x + y
```
and the first time the call site is encountered, both `x` and `y` are `int`s. The call site delegate will ask the language service, "For types `int` and `int`, whats the appropriate IL operation for `+` as an expression tree?" The language service returns a `BinaryExpression` with a `NodeType` of `Add` and the appropriate expressions at the `Left` and `Right` properties. The call site delegate then incorporates that expression tree into it's current tree and compiles it; which means the next time `x` and `y` are both `int`'s, the result will be calculated using a compiled method.

If the types change, e.g. now `x` is a `string`, the call site turns to the language service and gets a new expression tree. Rinse and repeat.

The `dynamic` features of C# make use of this mechanism.

There's more information about this at the [DLR GitHub repo](https://github.com/IronLanguages/dlr/tree/master/Docs). A good starting point is at Chapter 4 in the `dlr-overview.doc` or corresponding PDF,)

**Comment by binarycow** [comment]
@zspitz - I have since moved away from the expression extensions for this purpose - they were just too cumbersome to use.  (Though, I plan on using them for **other** things - namely, supporting a "foreach" construct, and then handling the `Reduce` method to reduce it to a standard `Expression.Loop`.)

No matter which method you decide to use, the important things for me, is that I can specify how to format the Extension nodes, without having to specify how to format the default nodes.  I also never use any formatter other than C#.

----

Thanks for the tip of the dynamic stuff.  I wasn't aware it the dynamic expression trees cached the compiled delegates.  That may be good enough!

**Comment by zspitz** [comment]
> Though, I plan on using them for **other** things - namely, supporting a "foreach" construct, and then handling the `Reduce` 

You may want to look at Bart de Smet's C# expression extensions, specifically, his [implementation of `foreach`](https://github.com/bartdesmet/ExpressionFutures/tree/master/CSharpExpressions#foreach).

**Comment by binarycow** [comment]
Thanks for the heads-up on that too.  I've got some extension nodes that wouldn't be covered by those, though.  So, maybe a mixture of both!

**Comment by zspitz** [comment]
@binarycow I've described the extensibility story [here](https://github.com/zspitz/ExpressionTreeToString/wiki/Extensibility-and-plugin-mechanism). Could you look over it when you have a chance, and see if it makes sense? Once you do so, I can close this issue.

**Comment by zspitz** [comment]
I'm closing this now. If you see a need to reopen, please do so.
